hail/src/main/scala/is/hail/expr/types/physical/PCanonicalArray.scala

```diff
@@ -367,4 +386,140 @@ final case class PCanonicalArray(elementType: PType, required: Boolean = false)
+    val numberOfElements = mb.newLocal[Int]
+    val currentElementAddress = mb.newLocal[Long]
+    val currentIdx = mb.newLocal[Int]
+
+    var c = Code(
+      numberOfElements := sourceType.loadLength(srcAddress),
+      dstAddress := this.allocate(region, numberOfElements),
+      this.stagedInitialize(dstAddress, numberOfElements),
+      currentElementAddress := this.firstElementOffset(dstAddress, numberOfElements),
+      currentIdx := const(0)
+    )
+
+    var loop: Code[Unit] =
+      destElementType.storeShallowAtOffset(
+        currentElementAddress,
+        destElementType.copyFromType(
+          mb,
+          region,
+          sourceElementType,
+          sourceType.loadElementAddress(srcAddress, numberOfElements, currentIdx),
+          allowDowncast,
+          forceDeep
+        )
+      )
+
+    if(destElementType.required > sourceElementType.required) {
+      assert(allowDowncast)
+
+      c = Code(sourceType.hasMissingValues(srcAddress).orEmpty(
+        Code._fatal("Found missing values. Cannot copy to type whose elements are required.")
+      ), c)
+    } else if(!sourceElementType.required) {
+      loop = sourceType.isElementMissing(srcAddress, currentIdx).mux(
+        this.setElementMissing(dstAddress, currentIdx),
+        loop
+      )
+    }
+
+    Code(
+      c,
+      Code.whileLoop(currentIdx < numberOfElements,
+        loop,
+        currentElementAddress := this.nextElementAddress(currentElementAddress),
+        currentIdx := currentIdx + const(1)
+      ),
+      dstAddress
+    )
+  }
+
+  override def copyFromType(region: Region, srcPType: PType, srcAddress: Long,
+    allowDowncast: Boolean, forceDeep: Boolean): Long = {
+    assert(srcPType.isInstanceOf[PArray])
+
+    val sourceType = srcPType.asInstanceOf[PArray]
+    val sourceElementType = sourceType.elementType.fundamentalType
+    val destElementType = this.elementType.fundamentalType
+
+    assert(sourceElementType isOfType destElementType)
+
+    if (sourceElementType == destElementType) {
+      if(!forceDeep) {
+        return srcAddress
+      }
+
+      if(sourceElementType.isPrimitive) {
+        return this.copyFrom(region, srcAddress)
+      }
+    }
+
+    if(destElementType.required > sourceElementType.required) {
+      assert(allowDowncast)
+
+      if(sourceType.hasMissingValues(srcAddress)) {
+        fatal("Found missing values. Cannot copy to type whose elements are required.")
+      }
+    }
+
+    val numberOfElements = sourceType.loadLength(srcAddress)
+    val dstAddress = this.allocate(region, numberOfElements)
+    this.initialize(region, dstAddress, numberOfElements)
+
+    var currentElementAddress = this.firstElementOffset(dstAddress, numberOfElements)
+    var currentIdx = 0
+
+    while(currentIdx < numberOfElements) {
+      if(!sourceElementType.required && sourceType.isElementMissing(region, srcAddress, currentIdx)) {
+        this.setElementMissing(dstAddress, currentIdx)
```

Comment:
unfortunately, calling this without `region` as the first argument, caused the staged version to be dispatched, due to implicit conversion!

Reply:
ack!! those implicit conversions...